pkg/compactor/partitioned_group_info.go

```diff
@@ -0,0 +1,307 @@
+}
+
+func (p *PartitionedGroupInfo) getPartitionIDsByBlock(blockID ulid.ULID) []int {
+	var partitionIDs []int
+partitionLoop:
+	for _, partition := range p.Partitions {
+		for _, block := range partition.Blocks {
+			if block == blockID {
+				partitionIDs = append(partitionIDs, partition.PartitionID)
+				continue partitionLoop
+			}
+		}
+	}
+	return partitionIDs
+}
+
+func (p *PartitionedGroupInfo) getAllBlocks() []ulid.ULID {
+	uniqueBlocks := make(map[ulid.ULID]struct{})
+	for _, partition := range p.Partitions {
+		for _, block := range partition.Blocks {
+			uniqueBlocks[block] = struct{}{}
+		}
+	}
+	blocks := make([]ulid.ULID, len(uniqueBlocks))
+	i := 0
+	for block := range uniqueBlocks {
+		blocks[i] = block
+		i++
+	}
+	return blocks
+}
+
+func (p *PartitionedGroupInfo) getPartitionedGroupStatus(
+	ctx context.Context,
+	userBucket objstore.InstrumentedBucket,
+	partitionVisitMarkerTimeout time.Duration,
+	userLogger log.Logger,
+) PartitionedGroupStatus {
+	status := PartitionedGroupStatus{
+		PartitionedGroupID:        p.PartitionedGroupID,
+		CanDelete:                 false,
+		IsCompleted:               false,
+		DeleteVisitMarker:         false,
+		PendingPartitions:         0,
+		InProgressPartitions:      0,
+		PendingOrFailedPartitions: []Partition{},
+	}
+	allPartitionCompleted := true
+	hasInProgressPartitions := false
+	for _, partition := range p.Partitions {
+		partitionVisitMarker := &PartitionVisitMarker{
+			PartitionedGroupID: p.PartitionedGroupID,
+			PartitionID:        partition.PartitionID,
+		}
+		visitMarkerManager := NewVisitMarkerManager(userBucket, userLogger, "PartitionedGroupInfo.getPartitionedGroupStatus", partitionVisitMarker)
+		partitionVisitMarkerExists := true
+		if err := visitMarkerManager.ReadVisitMarker(ctx, partitionVisitMarker); err != nil {
+			if errors.Is(err, errorVisitMarkerNotFound) {
+				partitionVisitMarkerExists = false
+			} else {
+				level.Warn(userLogger).Log("msg", "unable to read partition visit marker", "path", partitionVisitMarker.GetVisitMarkerFilePath(), "err", err)
+				return status
+			}
+		}
+
+		if !partitionVisitMarkerExists {
+			status.PendingPartitions++
+			allPartitionCompleted = false
+			status.PendingOrFailedPartitions = append(status.PendingOrFailedPartitions, partition)
+			continue
+		}
+
+		if partitionVisitMarker.VisitTime < p.CreationTime {
+			status.DeleteVisitMarker = true
+			allPartitionCompleted = false
+			continue
+		}
+
+		if partitionVisitMarker.GetStatus() == Completed {
+			continue
+		}
+
+		if (partitionVisitMarker.GetStatus() == Pending || partitionVisitMarker.GetStatus() == InProgress) && !partitionVisitMarker.IsExpired(partitionVisitMarkerTimeout) {
+			status.InProgressPartitions++
+			hasInProgressPartitions = true
+			allPartitionCompleted = false
+			continue
+		}
+
+		if partitionVisitMarker.GetStatus() == Failed {
+			status.PendingPartitions++
+		} else {
+			status.PendingPartitions++
```

Comment:
There is something off here. I think we can remove this else part

Reply:
Basically, we don't need to check `Failed` here. When code reaching here, the partition is either failed or not complete with expired visit marker. In both cases, the partition should be ready to be picked up for compaction.

I refactored the code to make it a bit more clear.